### PR TITLE
[MLIR][LLVM] DI Expression Rewrite & Legalization

### PR DESCRIPTION
**Pull request by zyx-billy**
Add a rewriter for DIExpressions & use it to run legalization patterns before exporting to llvm (because LLVM dialect allows DI Expressions that may not be valid in LLVM IR).

The rewriter driver works similarly to the existing mlir rewriter drivers, except it operates on lists of DIExpressionElemAttr (i.e. DIExpressionAttr). Each rewrite pattern transforms a range of DIExpressionElemAttr into a new list of DIExpressionElemAttr. 

In addition, this PR sets up a place to add legalization patterns that are broadly applicable internally to the LLVM dialect, and they will always be applied prior to export. This PR adds one pattern for merging fragment operators.

### CODE REVIEW

**Comment by Mogball** [comment]
Awesome! Please clarity in the motivation that the LLVM dialect allows DI expression to be valid that may not be valid in LLVM, hence why we need the legalization pass

File: mlir/include/mlir/Dialect/LLVMIR/Transforms/DIExpressionRewriter.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  using OperatorT = LLVM::DIExpressionElemAttr;
+
+  class ExprRewritePattern {
+  public:
+    using OperatorT = DIExpressionRewriter::OperatorT;
+    using OpIterT = std::deque<OperatorT>::const_iterator;
+    using OpIterRange = llvm::iterator_range<OpIterT>;
+
+    virtual ~ExprRewritePattern() = default;
+    /// Check whether a particular prefix of operators matches this pattern.
```
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/Transforms/DIExpressionRewriter.h:38]:
```suggestion
    /// Checks whether a particular prefix of operators matches this pattern.
```
ultra nit: usually we use third person plural for the first verb of a comment (here and below)

File: mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionLegalization.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+//===----------------------------------------------------------------------===//
+// Runner
+//===----------------------------------------------------------------------===//
+
+void mlir::LLVM::legalizeDIExpressionsRecursively(Operation *op) {
+  LLVM::DIExpressionRewriter rewriter;
+  rewriter.addPattern(std::make_unique<MergeFragments>());
+
+  mlir::AttrTypeReplacer replacer;
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionLegalization.cpp:56]:
```suggestion
  AttrTypeReplacer replacer;
```
ultra nit:

File: mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  // Invariant: concat(result, inputs) is equivalent to `operators` after some
+  // application of the rewrite patterns.
+  // Using a deque for inputs so that we have efficient front insertion and
+  // removal. Random access is not necessary for patterns.
+  std::deque<OperatorT> inputs(operators.begin(), operators.end());
+  SmallVector<OperatorT> result;
+
+  int64_t numRewrites = 0;
+  while (!inputs.empty() &&
+         (maxNumRewrites < 0 || numRewrites < maxNumRewrites)) {
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp:42]:
```suggestion
         (maxNumRewrites < 0 || numRewrites < maxNumRewrites)) {
```
Wouldn't std::optional<uint64_t> be much nicer than using -1 in this case?

**Review by gysit — APPROVED** [review]
Nice! 

LGTM modulo some minor comments.

File: mlir/include/mlir/Dialect/LLVMIR/Transforms/DIExpressionLegalization.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#include "mlir/Dialect/LLVMIR/Transforms/DIExpressionRewriter.h"
+
+namespace mlir {
+namespace LLVM {
+
+//===----------------------------------------------------------------------===//
+// Rewrite Patterns
+//===----------------------------------------------------------------------===//
+
+/// Adjacent DW_OP_LLVM_fragment should be merged into one.
```
Comment (Mogball) [inline mlir/include/mlir/Dialect/LLVMIR/Transforms/DIExpressionLegalization.h:26]: suggestion: please elaborate with a short psuedocode example, like `expr(frag, frag) -> expr(frag)` or something.

File: mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp (hunk 3 above)
Comment (Mogball) [reply mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp:42]: +1

**Review by Mogball — APPROVED** [review]
Thanks!

File: mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp (hunk 3 above)
Comment (zyx-billy) [reply mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp:42]: sounds good 👍 was a bit hesitant to wrap an int since the mlir rewriter was not, but I like this 😂 .

File: mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    if (!foundMatch) {
+      // If no match, pass along the current operator.
+      result.push_back(inputs.front());
+      inputs.pop_front();
+    }
+  }
+
+  if (maxNumRewrites.has_value() && numRewrites >= maxNumRewrites.value()) {
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp:66]:
```suggestion
  if (maxNumRewrites && numRewrites >= *maxNumRewrites) {
```
Drive-by comment :). The optional evaluates to a boolean so you can drop has_value() and if you want to access the value you can use the star syntax *maxNumRewrites. I think MLIR prefers these short-forms.

Comment (zyx-billy) [reply mlir/lib/Dialect/LLVMIR/Transforms/DIExpressionRewriter.cpp:66]: Oh awesome. Just in time. Thanks 🙏
